networkoperator/v1/types_egressrouter.go

```diff
@@ -0,0 +1,274 @@
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// Specification of the desired egress router.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinProperties:=4
+	// +kubebuilder:validation:MaxProperties:=4
+	Spec EgressRouterSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
+
+	// Observed status of EgressRouter.
+	// +kubebuilder:validation:Optional
+	// +optional
+	Status EgressRouterStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+
+// EgressRouterSpec contains the configuration for an egress router.
+// Mode, networkInterface and addresses fields must be specified along with exactly one "Config" that matches the mode.
+// Each config consists of parameters specific to that mode.
+// +k8s:openapi-gen=true
+// +kubebuilder:validation:Required
+type EgressRouterSpec struct {
+	// Mode depicts the mode that is used for the egress router. The default mode is "Redirect" and is the only supported mode currently.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Enum="Redirect"
+	// +kubebuilder:default:="Redirect"
+	Mode EgressRouterMode `json:"mode" protobuf:"bytes,1,opt,name=mode,casttype=EgressRouterMode"`
+
+	// Redirect represents the configuration parameters specific to redirect mode.
+	// +kubebuilder:validation:Optional
+	Redirect *RedirectConfig `json:"redirect,omitempty" protobuf:"bytes,2,opt,name=redirect"`
+
+	// Specification of interface to create/use. The default is macvlan.
+	// Currently only macvlan is supported.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinProperties:=1
+	// +kubebuilder:validation:MaxProperties:=1
+	// +kubebuilder:default:={macvlan: {mode: Bridge}}
+	NetworkInterface EgressRouterInterface `json:"networkInterface" protobuf:"bytes,3,opt,name=networkInterface"`
+
+	// List of IP addresses to configure on the pod's secondary interface.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinItems:=1
+	Addresses []EgressRouterAddress `json:"addresses" protobuf:"bytes,4,rep,name=addresses"`
+}
+
+// EgressRouterMode defines the different types of modes that are supported for the egress router interface.
+// The default mode is "Redirect" and is the only supported mode currently.
+type EgressRouterMode string
+
+const (
+	// EgressRouterModeRedirect creates an egress router that sets up iptables rules to redirect traffic
+	// from its own IP address to one or more remote destination IP addresses.
+	EgressRouterModeRedirect EgressRouterMode = "Redirect"
+)
+
+// RedirectConfig represents the configuration parameters specific to redirect mode.
+type RedirectConfig struct {
+	// List of L4RedirectRules that define the DNAT redirection from the pod to the destination in redirect mode.
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:MinItems:=1
+	RedirectRules []L4RedirectRule `json:"redirectRules,omitempty" protobuf:"bytes,1,rep,name=redirectRules"`
```

Comment:
is empty list and nil mean something different? Optional and minItems=1 suggest that.

Reply:
not sure I follow, what I meant here was that I want redirectrules array as an optional field, however if you do set it, then you need to provide at least one row in the array.

Reply:
That's what I mean: undefined is allowed, empty is not. Usually it's a good rule of thumb that undefined and empty mean the same. Either both are valid or none.

Reply:
yeah, sttts is right - get rid of MinItems I think?

Reply:
done. removed minItems.